daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PState.scala

```diff
@@ -289,7 +297,10 @@ final class PState private (
    * just to track that the variable was read in case we need to backtrack.
    */
   def variableRead(vrd: VariableRuntimeData) {
-    changedVariablesStack.top += vrd.globalQName
+    val maybeVar = variableMap.find(vrd.globalQName)
+    Assert.invariant(maybeVar.isDefined)
+    if (maybeVar.get.state != VariableRead)
+        changedVariablesStack.top += vrd.globalQName
```

Comment:
This means that there's now strict ordering between readVariable and variableRead (confusingly named, maybe a namechange is due?), correct? This function must always be called before readVariable, otherwise readVariable will change the state to VariableRead and then this will never add anything to the stack? This feels  a bit fragile--I could easily see someone changing the order for some reason and then things break. Is there maybe a way to call variableRead from readVariable?

Reply:
Yes, there is strict ordering for this as currently implemented. Having readVariable be managed by DState while setVariable and newVariableInstance are handled by PState makes things a bit ugly here. Would it perhaps be better to implement a corresponding readVariable call in PState and UState and have DState.VRef() call that instead?

Reply:
Yeah, I think that would clean things up a bit. Thoughts on calling the PState/UState function from readVariable, since readVariable actually knows when variable state changes to VariableRead, since a read doesnt always modify that state. That would avoid this VariableRead check.

Reply:
When these sorts of ordering things come up, scaladoc helps, but I always try to put in some sort of a Assert.precondition() that is true only if things are being called in the right order.

Reply:
Ok, will do.

Reply:
The problem with calling PState from VariableMap.readVariable() is that it doesn't know anything about PState. I thought we might be able to do a referringContext.isInstanceOf[PState], but it is always either a DPathCompileInfo or DPathElementCompileInfo. I think the call will need to remain strictly ordered and from VRef()

Reply:
I don't feel strongly one way or the other, but you could modify readVariable to take in ParseOrUnparseState as a parameter, which VRef has access to from the DState. I'm not sure if that's any better or worse than other options, but that's the way to do it for that option.

Reply:
Not sure why I didn' t think of that. This strange connection between DState and PState with the variableMap is confusing.